amazon/ion/simpleion.py

```diff
@@ -453,3 +459,64 @@ def loads(ion_str, catalog=None, single_value=True, encoding='utf-8', cls=None,
     return load(ion_buffer, catalog=catalog, single_value=single_value, encoding=encoding, cls=cls,
                 object_hook=object_hook, parse_float=parse_float, parse_int=parse_int, parse_constant=parse_constant,
                 object_pairs_hook=object_pairs_hook, use_decimal=use_decimal)
+
+
+def dump_extension(obj, fp, binary=True, sequence_as_stream=False, tuple_as_sexp=False, omit_version_marker=False):
+    res = ionc.ionc_write(obj, binary, sequence_as_stream, tuple_as_sexp)
+
+    # TODO support "omit_version_marker" rather than hacking.
+    if not binary and not omit_version_marker:
+        res = b'$ion_1_0 ' + res
+    fp.write(res)
+
+
+def load_extension(fp, single_value=True, encoding='utf-8'):
+    data = fp.read()
+    data = data if isinstance(data, bytes) else bytes(data, encoding)
+    return ionc.ionc_read(data, single_value, False)
+
+
+def dump(obj, fp, imports=None, binary=True, sequence_as_stream=False, skipkeys=False, ensure_ascii=True,
+         check_circular=True, allow_nan=True, cls=None, indent=None, separators=None, encoding='utf-8', default=None,
+         use_decimal=True, namedtuple_as_object=True, tuple_as_array=True, bigint_as_string=False, sort_keys=False,
+         item_sort_key=None, for_json=None, ignore_nan=False, int_as_string_bitcount=None, iterable_as_array=False,
+         tuple_as_sexp=False, omit_version_marker=False, **kw):
+    if c_ext and imports is None:
+        try:
+            return dump_extension(obj, fp, binary=binary, sequence_as_stream=sequence_as_stream,
+                              tuple_as_sexp=tuple_as_sexp, omit_version_marker=omit_version_marker)
+        except:
```

Comment:
Due to beta stage, I used above try...except to handle c extension unexpected issue. If there is some issue for c extension we run the original python API instead.

This try...catch statement will be eventually removed after c extension becomes more stable.

Reply:
Is it possible for dump to fail after partially writing to `fp`? If so, this attempt at recovery won't really work.

Reply:
I removed the try..catch branch. 

Instead of some recovery logic, we may just put c extension in the beta stage for a while until it's more stable, is it sounds good?

One of the reasons I tried to put it in the beta stage is because some CPython layer issue such as wrong type conversion will crash python instead of throwing exceptions.

Reply:
Yeah, let's just rely on beta testing.